File: CollectionLayout/ViewController.m
//
//  ViewController.m
//  CollectionLayout
//
//  Created by 苏沫离 on 2020/8/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//

#import "ViewController.h"
#import "CollectionAlignmentCell.h"

@interface UIViewController (Data)
- (void)reloadData:(NSMutableArray<DataModel *> *)dataArray;

@end

@implementation UIViewController (Data)
- (void)reloadData:(NSMutableArray<DataModel *> *)dataArray{}
@end


@interface ViewController ()
<UICollectionViewDelegate,UICollectionViewDataSource>
@property (nonatomic ,strong) UICollectionView *collectionView;
@property (nonatomic ,strong) NSMutableArray<NSDictionary<NSString * ,NSString *> *> *itemArray;
@property (nonatomic ,strong) NSMutableArray<DataModel *> *dataArray;
@end

@implementation ViewController

#pragma mark - life cycyle

- (void)viewDidLoad {
    [super viewDidLoad];
    self.navigationItem.title = @"Flow Layout";
    
    [self.view addSubview:self.collectionView];
    [DataModel creatDemoData:^(NSMutableArray<DataModel *> * _Nonnull array) {
        self.dataArray = array;
    }];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - UICollectionViewDelegate

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.itemArray.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    CollectionAlignmentCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass(CollectionAlignmentCell.class) forIndexPath:indexPath];
    cell.itemLable.text = self.itemArray[indexPath.row].allKeys.firstObject;
    return cell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    UIViewController *vc = [[NSClassFromString(self.itemArray[indexPath.row].allValues.firstObject) alloc] init];
    vc.navigationItem.title = self.itemArray[indexPath.row].allKeys.firstObject;
    [self.navigationController pushViewController:vc animated:YES];
    [vc reloadData:self.dataArray];
}

#pragma mark - setter and getter

- (NSMutableArray<NSDictionary<NSString *,NSString *> *> *)itemArray{
    if(_itemArray == nil){
        _itemArray = [NSMutableArray array];
        [_itemArray addObject:@{@"左对齐/右对齐":@"AlignmentViewController"}];
        [_itemArray addObject:@{@"瀑布流":@"PinterestViewController"}];
        [_itemArray addObject:@{@"卡片覆盖效果":@"CardViewController"}];
        [_itemArray addObject:@{@"卡片轮转效果":@"CardReelViewController"}];
        [_itemArray addObject:@{@"过渡动画":@"TransitionAnimationViewController"}];
    }
    return _itemArray;
}

- (UICollectionView *)collectionView{
    if (_collectionView == nil){
        UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
        flowLayout.minimumLineSpacing = 16;//垂直间距
        flowLayout.itemSize = CGSizeMake(CGRectGetWidth(self.view.bounds) - 24, 44);
        
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0,CGRectGetWidth(self.view.frame), CGRectGetHeight(self.view.frame)) collectionViewLayout:flowLayout];
        _collectionView.contentInset = UIEdgeInsetsMake(12, 12, 12, 12);
        _collectionView.showsVerticalScrollIndicator = NO;
        _collectionView.showsHorizontalScrollIndicator = NO;
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        _collectionView.backgroundColor = [UIColor whiteColor];
        [_collectionView registerClass:[CollectionAlignmentCell class] forCellWithReuseIdentifier:NSStringFromClass(CollectionAlignmentCell.class)];
    }
    return _collectionView;
}

@end

File: CollectionLayout-Swift/CollectionLayout-Swift/ControllerGroup/ViewController.swift
//
//  ViewController.swift
//  CollectionLayout-Swift
//
//  Created by 苏沫离 on 2020/9/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UICollectionViewDelegate,UICollectionViewDataSource {
    var collectionView : UICollectionView!
    var dataArray : NSMutableArray!
    let itemArray : NSMutableArray = NSMutableArray()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.white
        self.navigationItem.title = "Flow Layout"
        
         itemArray.add(["左对齐/右对齐":"AlignmentViewController"])
         itemArray.add(["瀑布流":"PinterestViewController"])
         itemArray.add(["卡片覆盖效果":"CardViewController"])
         itemArray.add(["卡片轮转效果":"CardReelViewController"])
         itemArray.add(["过渡动画":"TransitionAnimationViewController"])
        
        
        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.minimumLineSpacing = 16
        flowLayout.itemSize = CGSize(width: self.view.bounds.width - 24, height: 44)
        
        collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: flowLayout)
        collectionView.contentInset = UIEdgeInsets.init(top: 12, left: 12, bottom: 12, right: 12)
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.delegate = self
        collectionView.dataSource = self
        collectionView.backgroundColor = UIColor.white
        collectionView.register(CollectionAlignmentCell.self, forCellWithReuseIdentifier: "CollectionAlignmentCell")
        self.view.addSubview(collectionView)
        DataModel.shareDemoData()
        
        print(self.view)
        print(UIScreen.main)
    }

    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return itemArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell : CollectionAlignmentCell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionAlignmentCell", for: indexPath) as! CollectionAlignmentCell
        let dict : NSDictionary = itemArray.object(at: indexPath.row) as! NSDictionary
        cell.itemLable.text = (dict.allKeys.first as! String)
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let dict : NSDictionary = itemArray.object(at: indexPath.row) as! NSDictionary
        let key = dict.allKeys.first as! String
        
        if key == "左对齐/右对齐" {
            let vc : AlignmentViewController = AlignmentViewController()
            vc.navigationItem.title = key
            self.navigationController?.pushViewController(vc, animated: true)
        }else if key == "卡片轮转效果" {
            let vc : CardReelViewController = CardReelViewController()
            vc.navigationItem.title = key
            self.navigationController?.pushViewController(vc, animated: true)
        }else if key == "卡片覆盖效果" {
            let vc : CardViewController = CardViewController()
            vc.navigationItem.title = key
            self.navigationController?.pushViewController(vc, animated: true)
        }else if key == "瀑布流" {
            let vc : PinterestViewController = PinterestViewController()
            vc.navigationItem.title = key
            self.navigationController?.pushViewController(vc, animated: true)
        }else if key == "过渡动画" {
            let vc : TransitionAnimationViewController = TransitionAnimationViewController()
            vc.navigationItem.title = key
            self.navigationController?.pushViewController(vc, animated: true)
        }
    }
}

